### PR TITLE
MULE-12434: Poll extension model is invalid

### PR DESCRIPTION
**Pull request by pablomusumeci**

### CODE REVIEW

File: src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java [hunk 1]
```diff
@@ -38,25 +40,34 @@
 public final class DefaultTypeCatalog implements TypeCatalog {
 
   private List<SubTypesMappingContainer> mappings = new LinkedList<>();
-  private Map<String, ObjectType> types = new LinkedHashMap<>();
+  private Table<String, String, ObjectType> types = TreeBasedTable.create();
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java]: add a comment explaining what each value is

Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java]: I think this is better suited by a Map<String, Pair>... I would rather move Pair to api and use here than adding guava just for this

Comment (pablomusumeci) [inline src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java]: This is more like a Map<String, Map<String, ObjectType>>

File: pom.xml [hunk 2]
```diff
@@ -145,6 +146,11 @@
             <artifactId>gson</artifactId>
             <version>${gsonVersion}</version>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
```
Comment (alegmarra) [inline pom.xml]: have you built the distros/updated the assembly?

File: src/main/java/org/mule/runtime/api/meta/type/TypeCatalog.java [hunk 3]
```diff
@@ -93,4 +93,14 @@ static TypeCatalog getDefault(Set<ExtensionModel> extensions) {
...
 
+  /**
+   * @return a {@link Collection} with all the types declared by the extension with name {@code extensionName}
+   */
+  Collection<ObjectType> getExtensionTypes(String extensionName);
+
+  /**
+   * @return an {@link Optional} with the extension name that contributed with this given type
+   */
+  Optional<String> getExtension(String typeId);
```
Comment (alegmarra) [inline src/main/java/org/mule/runtime/api/meta/type/TypeCatalog.java]: getDeclaringExtension

File: src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java [hunk 4]
```diff
@@ -38,25 +40,34 @@
...
       mappings.add(new SubTypesMappingContainer(e.getSubTypes()));
-      e.getTypes().forEach(t -> getTypeId(t).ifPresent(id -> types.put(id, t)));
+      e.getTypes().forEach(t -> getTypeId(t).ifPresent(id -> types.put(e.getName(), id, t)));
     });
   }
 
   @Override
   public Optional<ObjectType> getType(String typeId) {
-    return ofNullable(types.get(typeId));
+    return types.column(typeId).values().stream().findFirst();
```
Comment (alegmarra) [inline src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java]: starting to wonder if is not better just to have two indexed structures, <typeId, type>; <typeId, extension> (not extensionName, the actual extension)

Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java]: the answer to that depends on how and how often is this structure queried

Comment (alegmarra) [inline src/main/java/org/mule/runtime/internal/meta/type/DefaultTypeCatalog.java]: lots

**Pull request merged**